### PR TITLE
support @extends in jsdoc

### PR DESCRIPTION
**Pull request by aozgaa**
Fixes https://github.com/Microsoft/TypeScript/issues/18701

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -4765,8 +4765,8 @@ namespace ts {
         return node.kind === SyntaxKind.JSDocComment;
     }
 
-    export function isJSDocAugmentsTag(node: Node): node is JSDocAugmentsTag {
-        return node.kind === SyntaxKind.JSDocAugmentsTag;
+    export function isJSDocAugmentsOrExtendsTag(node: Node): node is JSDocAugmentsOrExtendsTag {
```
Comment (mhegazy) [inline src/compiler/utilities.ts]: this is part of the public API now, we should not break that.

**Comment by andy-ms** [comment]
Looks like this will cause merge conflicts in #18775, but I'll handle it.

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -423,8 +423,9 @@ namespace ts {
                 return visitNode(cbNode, (<JSDocReturnTag>node).typeExpression);
             case SyntaxKind.JSDocTypeTag:
                 return visitNode(cbNode, (<JSDocTypeTag>node).typeExpression);
-            case SyntaxKind.JSDocAugmentsTag:
-                return visitNode(cbNode, (<JSDocAugmentsTag>node).typeExpression);
+            case SyntaxKind.JSDocAugmentsOrExtendsTag:
+            case SyntaxKind.JSDocExtendsTag:
```
Comment (weswigham) [inline src/compiler/parser.ts]: It seems like `SyntaxKind.JSDocExtendsTag` is never assigned anywhere?

File: src/compiler/types.ts [hunk 3]
```diff
@@ -363,7 +363,8 @@ namespace ts {
         JSDocVariadicType,
         JSDocComment,
         JSDocTag,
-        JSDocAugmentsTag,
+        JSDocAugmentsOrExtendsTag,
```
Comment (sandersn) [inline src/compiler/types.ts]: I think we should keep `JSDocAugmentsTag` and just let `@extends` be a silent synonym for it in the parser. I don't think it's bad, for example, if the roundtrip emit converts `@extends` to `@augment`. And that's the worst side-effect I can think of.

**Comment by andy-ms** [comment]
@sandersn mentioned in https://github.com/Microsoft/TypeScript/pull/18854#pullrequestreview-66523402 that we should have an error if both `@extends` and `@augments` tags are present, but continue on using just the type from `@extends`.

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -4247,6 +4247,12 @@ namespace ts {
         return find(tags, doc => doc.kind === kind);
     }
 
+    /** Gets all JSDoc tags of a specified kind, or undefined if not present. */
+    export function getAllJSDocTagsOfKind(node: Node, kind: SyntaxKind): ReadonlyArray<JSDocTag> | undefined {
```
Comment (sandersn) [inline src/compiler/utilities.ts]: remove 'All'

**Pull request merged**